### PR TITLE
adds FreeBSD platform

### PR DESCRIPTION
**Pull request by ericoporto**
I ended up adding a platform driver named Generic Unix, but I specialize it for FreeBSD so I can have a `eOSFreeBSD`. I am not sure if this is the best approach, in my head it makes easier to add more platforms. 

I only used CMake since it's written to build the most portable binary as possible - so it's better for actually shipping binary games.

For system packages is better that the person packaging it writes their own build aware of whatever packages are available in the specific FreeBSD version. Unfortunately no packager ever returned any of my e-mails, so this may take a while.

Still, the system specifics defines in headers are now alright for FreeBSD.

I don't plan to provide CI builds for FreeBSD.

### CODE REVIEW

File: Common/core/platform.h [hunk 1]
```diff
@@ -32,36 +34,48 @@
     #endif
 
     #if TARGET_OS_SIMULATOR || TARGET_IPHONE_SIMULATOR
-        #define AGS_PLATFORM_OS_WINDOWS (0)
```
Comment (rofl0r) [inline Common/core/platform.h:35]: i don't know about AGS stance on this, but in other projects i contribute too the general consensus is to not add gratuitous whitespace changes to unrelated changes.

Comment (ericoporto) [reply Common/core/platform.h:35]: They are not gratuitous, I will need the additional space soon and I don't know yet when this will be merged.

Comment (ivan-mogilko) [reply Common/core/platform.h:35]: I don't quite understand though, what are these spaces for?

**Comment by ivan-mogilko** [comment]
If I understand correctly, this should resolve #1265 ? Wanted to reference it here just in case.

File: Common/core/platform.h (hunk 1 above)
Comment (ericoporto) [reply Common/core/platform.h:35]: I just wanted to avoid merge conflicts with the Emscripten change in case this takes a while - it may not be trivial to setup a FreeBSD VM.

**Comment by ericoporto** [comment]
Yes, but not fully, the user clearly has an expectation to be able to download from ports but I haven't been able to reach any of the people that did the commits from that port. See the commit history below:

https://www.freshports.org/games/ags

Ideally someone would need to contact @mat813 or @t6 which are the maintainers (at least I think, neither answered e-mails).

File: Common/core/platform.h (hunk 1 above)
Comment (ivan-mogilko) [reply Common/core/platform.h:35]: My point was, you do not have to do this in Emscripten pr either, there's no real need to adjust the lengths of all the lines if you add a new one which is slightly longer.

**Comment by ivan-mogilko** [comment]
The biggest question I have in regards to this change is, what is the difference between AGSGenericUnix and AGSLinux platform driver? May it make sense to make AGSLinux also a child of AGSGenericUnix?

**Comment by ericoporto** [comment]
So, this is what I was most curious, right now there's no difference, and I think it could, except that the old CD support (does someone actually ever used that?) actually is there on Linux, but I have removed it from Generic Unix - I doubt anyone will ever test and verify, when I imagine sbc like Dragonfly picking it up or NetBSD.

Other thing is I temporarily left XDG support there (because it IS available on FreeBSD as far as I can tell), but I don't think XDG is something in all Unixes, only those that adhere to Freedesktop. 

So in my head it's like:

base (which could use SDL stuff) -> Generic/Posix Unix -> Unix with Freedesktop -> GNU/Linux

File: Common/core/platform.h (hunk 1 above)
Comment (ericoporto) [reply Common/core/platform.h:35]: I will rebase and these will not be in this PR.

**Comment by rofl0r** [comment]
i'd model it like so:
basically there's windows vs posix  (which means nothing else than "generic unix") code, and a os selection which turns on or off via macro certain platform specific code

e.g.
```c
#ifdef __linux__
#define OS_LINUX
#define PLATFORM_POSIX
#define PLATFORM_SUPPORTS_CDA
#define PLATFORM_SUPPORTS_XDG
#endif
```
etc.

> Other thing is I temporarily left XDG support there (because it IS available on FreeBSD as far as I can tell), but I don't think XDG is something in all Unixes, only those that adhere to Freedesktop (may need to be moved around).

tbh not even linux is so clear-cut, basically "linux" means nothing more than "uses linux kernel" but everything else can be different for every single distribution.

**Comment by ericoporto** [comment]
@rofl0r do you know if there's a way to detect XDG support? Better if at compiler time but alright if runtime, so we could something like, use these folders but if can't use these?

Also, should I separate the generic_unix from the tiny FreeBSD code? My mentality on leaving together is because the macros are defining what will build and what won't be built, so one could lump it on together when building and the compiler will select what makes sense.

I think I can add the `PLATFORM_SUPPORTS_CDA` directly on core/platform.h too. I don't like extremely long files but there it really saves time finding things together - I even thought on moving the library loading selection there but left out because it has 3 states instead of 2.

**Comment by ivan-mogilko** [comment]
> I think I can add the PLATFORM_SUPPORTS_CDA directly on core/platform.h too. I don't like extremely long files but there it really saves time finding things together

I don't think it's too long at the moment, but in the worst case we may split it into 2, where one has platform detection and second has options defined depending on the platform.

**Comment by ericoporto** [comment]
Found something interesting, SDL simply checks if a HOME environment variable exists (https://github.com/libsdl-org/SDL/blob/main/src/filesystem/unix/SDL_sysfilesystem.c)

I am thinking on having Generic Unix without libcda, and have it do the XDG test but slightly better at fallback and then have both Linux and FreeBSD build on it, switching the OS enum, and have Linux have libcda.

File: Engine/CMakeLists.txt [hunk 2]
```diff
@@ -419,6 +419,8 @@ target_sources(engine
     media/video/video.h
     platform/base/agsplatformdriver.cpp
     platform/base/agsplatformdriver.h
+    platform/base/agsplatform_unix.cpp
+    platform/base/agsplatform_unix.h
     platform/base/sys_main.cpp
```
Comment (rofl0r) [inline Engine/CMakeLists.txt:424]: the cpp file would need to be added to Engine/Makefile too i guess ?

Comment (ericoporto) [reply Engine/CMakeLists.txt:424]: so, the base platform apparently has some magic that makes unnecessary, because platform/base is already added.

https://github.com/adventuregamestudio/ags/blob/811f6c126fcff1cc5a0db8b6e0a1b9fca2550235/Engine/Makefile-objs#L30

but check the commit below, that one would be needed if there was a bsd makefile (I think)

**Comment by ericoporto** [comment]
@rofl0r I am not particularly confident on the names I chose for things, do you think the hierarchy makes sense?

**Comment by rofl0r** [comment]
> @rofl0r I am not particularly confident on the names I chose for things, do you think the hierarchy makes sense?

well, it's not quite as i lined out but looks OKish so far. i only don't like the "freedesktop platform" bit, which makes it sound as if freedesktop is a hugely important thing, when this is really just about XDG home directory layout. i'd just call it something like "uses_xdg", "has_xdg", "with_xdg", "supports_xdg" or something like that.

> the base platform apparently has some magic that makes unnecessary, because platform/base is already added.

oh great, i forgot we use a glob there.

**Comment by ericoporto** [comment]
Uhm, maybe `AGS_PLATFORM_IS_XDG_UNIX` ?

**Comment by ivan-mogilko** [comment]
Since you have a "AGSPlatformUnix" there, why not call it just AGS_PLATFORM_UNIX?

There are no other subtypes there at the moment, and we don't know if we will have any in the future.

**Comment by ivan-mogilko** [comment]
I think [rofl0r's original suggestion](https://github.com/adventuregamestudio/ags/pull/1324#issuecomment-865115436) was to consider XDG to be rather an optional component rather something that defines whole platform type?

What is the purpose of this base class, which functionality it shares to all common children?

What if we just call it "desktop unix" as opposed to mobile (e.g. Android). And then you may toggle XDG support with additional macro inside the class if required.

**Comment by ivan-mogilko** [comment]
Sorry nevermind, I removed some of my previous comments. Apparently I misunderstood what XDG really is.

TBH personally I would just call it "desktop unix" for now, to distinct from mobile and as something that would be clear to most developers.
But if "XDG unix" is a more correct term for this, okay then.

**Comment by ericoporto** [comment]
Yes, I agree, this is why I didn't want the name to be *just* Unix, thought it would be misleading.

**Comment by ivan-mogilko** [comment]
@ericoporto if I understand correctly, this PR is practically ready?
One thing, could you also rename `AGSPlatformUnix` to have XDG in the name to match the platform definition?

**Comment by ericoporto** [comment]
it's a bit long but it's done :grimacing:
